include/wil/property.h

```diff
@@ -0,0 +1,226 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+
+#pragma once
+#ifndef __WIL_PROPERTY_INCLUDED
+#define __WIL_PROPERTY_INCLUDED
+
+namespace wil
+{
+
+template <typename T>
+struct read_only_property

```

Comment:
I fail to see the benefit of the `read_only_property` and `read_write_property` types. If I want a read only property, I'll declare a `const` public member. If I want a read-write property, I'll declare a non-`const` public member.

Additionally, since this does nothing to prevent assignment, I'm pretty sure something like this will work:
```c++
obj->readOnlyProperty = read_only_property<int>(42);
```

Reply:
Thanks for the feedback. This is to implement WinRT properties as C++/WinRT expects properties to work.

In C++/WinRT, for a readonly property you have to declare a data member + method. For a read/write property, you have to declare a data member + 2 methods. This rapidly adds up to a bunch of repetitive boilerplate code. The types in this PR encapsulate the boilerplate.

Reply:
I see. In that case, it should _definitely_ move to a C++/WinRT specific header.

But that said, something like `read_write_property` seems very risky as it does not provide any mechanism for synchronization.

Reply:
renamed the header. 
note that nothing prevents you from continuing to write properties the way you do today, this just provides a low-cost abstraction to do away with all the boilerplate for 90% of cases.

Reply:
I do agree with the overall feeling. Why not simply make a `propert<T>` type and to make it read-only have people write `const property<int> Foo;`

Reply:
First thing first: I think allowing assignment to `read_only_property` is intentional? To allow write in implementation type, while disallowing write in projected type. Please correct me if I'm wrong here.

If this observation is right, the problem is that it seems `const property<int>` isn't writable even in implementation type. It'd behave like a constant, whereas `read_only_property` actually is read-only in projected type, but writable in implementation type (classic `get; private set;` property in C#).

Consider the following table for a typical winrt property's getter and setter. Public means callable in projected type. Private means callable only in implementation type.

**Current implementation**
|getter|setter|property to use|
|--|--|--|
|public|public| `read_write_property` |
|public|private| `read_only_property` (allows mutation via assignment) |
|private|public| Not supported |
|private|private| Don't need wrapper |

**Suggested**
|getter|setter|property to use|
|--|--|--|
|public|public| `property` |
|public|private| Nothing (From what I read, the suggested `const property<T>` _likely_ doesn't allow any mutation, private or not) |
|private|public| Not supported |
|private|private| Don't need wrapper|

This makes `read_only_property` more useful than `const property<int>`.

Granted, we can make `const property<int>` allow write via assignment as well - but that seems more awkward as the write has to work around C++'s type system using `mutable` and what not, whereas `read_only_property`'s `read-only-ness` is implicit in its overloading of `operator()`. To put it the other way, we don't really want the property to be read-only in C++'s type system, but only read-only in projected type (aka `get; private set`).

Maybe we just need a better name to indicate the property is only `read-only` in WinRT? e.g. `public_get_private_set_property`? `winrt_read_only_property`?